File: typescript/src/paths.ts
// Minimal JSONPath-like traversal, mirroring python/src/agentreplay/paths.py.

type Token =
  | { kind: "key"; value: string }
  | { kind: "index"; value: number }
  | { kind: "wildcard" };

const TOKEN_RE = /([^.\[\]]+)|\[([0-9]+|\*)\]/g;

function tokenize(path: string): Token[] {
  let p = path;
  if (p.startsWith("$")) {
    p = p.slice(1);
    if (p.startsWith(".")) p = p.slice(1);
  }
  const tokens: Token[] = [];
  let pos = 0;
  while (pos < p.length) {
    if (p[pos] === ".") {
      pos++;
      continue;
    }
    TOKEN_RE.lastIndex = pos;
    const m = TOKEN_RE.exec(p);
    if (!m || m.index !== pos) {
      throw new Error(`invalid path token near "${p.slice(pos, pos + 10)}" in ${path}`);
    }
    if (m[1] !== undefined) tokens.push({ kind: "key", value: m[1] });
    else if (m[2] === "*") tokens.push({ kind: "wildcard" });
    else tokens.push({ kind: "index", value: parseInt(m[2]!, 10) });
    pos = TOKEN_RE.lastIndex;
  }
  return tokens;
}

export type Container = Record<string, unknown> | unknown[];
export type Hit = { container: Container; key: string | number };

export function* iterTargets(root: unknown, path: string): Iterable<Hit> {
  const tokens = tokenize(path);
  yield* walk(root, tokens, 0);
}

function* walk(node: unknown, tokens: Token[], i: number): Iterable<Hit> {
  if (i >= tokens.length) return;
  const tok = tokens[i];
  const last = i === tokens.length - 1;

  if (tok.kind === "key") {
    if (node === null || typeof node !== "object" || Array.isArray(node)) return;
    const obj = node as Record<string, unknown>;
    if (!(tok.value in obj)) return;
    if (last) yield { container: obj, key: tok.value };
    else yield* walk(obj[tok.value], tokens, i + 1);
  } else if (tok.kind === "index") {
    if (!Array.isArray(node)) return;
    if (tok.value >= node.length) return;
    if (last) yield { container: node, key: tok.value };
    else yield* walk(node[tok.value], tokens, i + 1);
  } else {
    if (Array.isArray(node)) {
      for (let idx = 0; idx < node.length; idx++) {
        if (last) yield { container: node, key: idx };
        else yield* walk(node[idx], tokens, i + 1);
      }
    } else if (node !== null && typeof node === "object") {
      const obj = node as Record<string, unknown>;
      for (const k of Object.keys(obj)) {
        if (last) yield { container: obj, key: k };
        else yield* walk(obj[k], tokens, i + 1);
      }
    }
  }
}

export function deleteAt(container: Container, key: string | number): void {
  if (Array.isArray(container)) {
    // Sentinel-fill for arrays to preserve length (mirrors Python).
    (container as unknown[])[key as number] = "<NORMALIZED>";
  } else {
    delete (container as Record<string, unknown>)[key as string];
  }
}

File: python/src/agentreplay/paths.py
"""Minimal JSONPath-like traversal used by redaction & normalization.

Supported syntax
----------------
- Dotted keys:            ``a.b.c``
- Numeric index:          ``a[0]``
- Wildcard index:         ``a[*]``  (all elements)
- Leading ``$`` allowed but optional
- ``**`` (recursive descent) is intentionally NOT supported in v0.1.

The traversal engine yields ``(container, key_or_index)`` pairs so callers can
mutate in place (delete, replace) without threading paths back through.
"""

from __future__ import annotations

import re
from typing import Any, Iterator, Tuple

_TOKEN_RE = re.compile(r"([^.\[\]]+)|\[([0-9]+|\*)\]")


def _tokenize(path: str) -> list[tuple[str, str]]:
    """Return list of (kind, value) tokens.

    kind in {"key", "index", "wildcard"}.
    """
    if path.startswith("$"):
        path = path[1:]
        if path.startswith("."):
            path = path[1:]
    tokens: list[tuple[str, str]] = []
    pos = 0
    while pos < len(path):
        if path[pos] == ".":
            pos += 1
            continue
        m = _TOKEN_RE.match(path, pos)
        if not m:
            raise ValueError(f"invalid path token near {path[pos:pos+10]!r} in {path!r}")
        if m.group(1) is not None:
            tokens.append(("key", m.group(1)))
        else:
            g = m.group(2)
            if g == "*":
                tokens.append(("wildcard", "*"))
            else:
                tokens.append(("index", g))
        pos = m.end()
    return tokens


def iter_targets(obj: Any, path: str) -> Iterator[Tuple[Any, Any]]:
    """Yield ``(container, key)`` pairs for every location matching ``path``.

    ``container[key]`` is the matched value; deleting via ``del container[key]``
    or reassigning via ``container[key] = ...`` is safe.
    """
    tokens = _tokenize(path)
    yield from _walk(obj, tokens, 0)


def _walk(node: Any, tokens: list[tuple[str, str]], i: int) -> Iterator[Tuple[Any, Any]]:
    if i >= len(tokens):
        return
    kind, val = tokens[i]
    is_last = i == len(tokens) - 1

    if kind == "key":
        if not isinstance(node, dict) or val not in node:
            return
        if is_last:
            yield node, val
        else:
            yield from _walk(node[val], tokens, i + 1)
    elif kind == "index":
        if not isinstance(node, list):
            return
        idx = int(val)
        if idx >= len(node):
            return
        if is_last:
            yield node, idx
        else:
            yield from _walk(node[idx], tokens, i + 1)
    else:  # wildcard
        if isinstance(node, list):
            for idx in range(len(node)):
                if is_last:
                    yield node, idx
                else:
                    yield from _walk(node[idx], tokens, i + 1)
        elif isinstance(node, dict):
            for k in list(node.keys()):
                if is_last:
                    yield node, k
                else:
                    yield from _walk(node[k], tokens, i + 1)
